### PR TITLE
Added some improvements

### PR DESCRIPTION
**Pull request by Aerue**
Hey i'm working on some improvements. Take a look and tell me !
- [x] Improve accuracy of zoomer to avoid smaller images than viewport
- [x] Add possibility to control maxZoom, minZoom and initialZoom (#47) - (#48) ?
- [x] Correct deepExtend

### CODE REVIEW

**Comment by thedustinsmith** [comment]
Thanks for the PR.  I'm fine with the first item.  

Why did deepExtend need correcting?

The second one referring to minZoom, maxZoom, and iniitalZoom is really what's causing me hesitation.  You're definitely not the first person to request control of these values, either.  So, that's why I'd like to take this opportunity to open a discussion.

I'll agree that our `maxZoom` property needs a lot of work.  Right now I think it's either 150% of the image width or 100% of the min width.  

The problem I have is setting these values in the constructor of croppie.  I don't think it makes sense there.  Shouldn't the min and max zoom be based on the image that you're binding to the croppie?  If you're binding a very small image, would you want to limit that maxZoom so you're not zooming in on a pixel-y image?  Do you know what size image you're going to bind to the croppie when you're initializing croppie?  Take the upload demo, for example, I think you'd want to adjust those properties after you know the size of the image.

There's a lot of different ways to go with the min/max zoom stuff.  I just want to discuss it before we choose a direction to go.  What're you thoughts on this?

**Comment by Aerue** [comment]
> Why did deepExtend need correcting?

If you set a new Croppie object and console.log it in _create method you will see that properties aren't properly copied from Croppie.defaults. Properties are copied when boundary is not set, but if you set boundary.height deepextend don't copy with property from defaults. My deepExtend fix that.

> The problem I have is setting these values in the constructor of croppie. I don't think it makes sense there. Shouldn't the min and max zoom be based on the image that you're binding to the croppie? If you're binding a very small image, would you want to limit that maxZoom so you're not zooming in on a pixel-y image? Do you know what size image you're going to bind to the croppie when you're initializing croppie? Take the upload demo, for example, I think you'd want to adjust those properties after you know the size of the image.

I have done that PR in order to manage my images upload. At the base i don't wanna zoom on image, only dezoom, for example my settings on init are : initialZoom: 1, minZoom: 0, maxZoom: 1.

> If you're binding a very small image, would you want to limit that maxZoom so you're not zooming in on a pixel-y image?

You talk here to me about Validation, Croppie don't handle that, so maybe add validation on init params to allow/disallow image width/height/types ? But for me, when you init a new croppie it is to create many image on the same width/height/ratio (i use the size viewport), so changing params on the fly is not a good idea to me, bcz i don't have any use of that, but that shoul'd be good to allow that if our image width/height/ratio/minZoom, ... is not the same depending on a form select option element for example.

**Comment by thedustinsmith** [comment]
> If you set a new Croppie object and console.log it in _create method you will see that properties aren't properly copied from Croppie.defaults.

I'm not seeing the problem that you're describing.  `deepExtend` seems to be working exactly how I'd expect it to work.  If I initialize a croppie with 
```
{
boundary: { width: 400, height: 325 },
showZoomer: false
}
```
Then log out `self.options` in `_create`, I get:
```
{
boundary: {
   height: 325,
   width: 400
},
customClass: "",
enableZoom: true,
mouseWheelZoom: true,
showZoomer: false,
viewport: {
    height: 100,
    type: "square",
    width: 100
}
```

That seems right to me.

> At the base i don't wanna zoom on image, only dezoom, for example my settings on init are : initialZoom: 1, minZoom: 0, maxZoom: 1.

This makes sense to me - I can see the desire to restrict zooming to only zoom out.  You're right, Croppie shouldn't be responsible for validating.  At least not right now.  Let me try to explain what I meant a little better...

For your use case, you only want users to zoom out images.  Let's say your viewport is 100x100.  Someone uploads an image that is 50x50.  The minZoom is going to be 2.0, because croppie requires that the image is at least as wide as the viewport.  Then, what croppie currently does, is set your maxZoom to 3.0.  I'm not saying I like the functionality it's currently performing.  What I'm trying to say is you passed in `min: 0, max: 1, initial: 1`, and those values will not apply to this smaller image.  

That's why it seems like we should be doing something with these zoom values when you're binding each image, I just don't know what.

**Comment by Aerue** [comment]
About deepExtend try to do the test but with only

    boundary: {width: 400}

You will see the following in the _create :

    {
    boundary: {
       width: 400
    },
    customClass: "",
    enableZoom: true,
    mouseWheelZoom: true,
    showZoomer: false,
    viewport: {
        height: 100,
        type: "square",
        width: 100
    }

The height params of boundary is missing.

What i'm doing with this PR it's to add params on image result not on image upload/binded, because that's we want it's image result on the same size/ratio. What's the interest of bindinf params on each image, do we have an interest of having many image with different ratio ?

**Comment by thedustinsmith** [comment]
@Aerue Sorry - I've dropped off the radar on this one.  Been busy with a lot of other stuff.  I cherry picked your deepExtend changes from this branch.  Good call on those.  I also modified the zoom decimal points to 4 (I couldn't cherry pick that due to other changes).

I'd like to continue the discussion on the zoom changes, but I'm not ready for those to be part of the stable release yet.  I realize that they won't break anything being in there, but I'd rather make a collaborative decision on how those parameters should be handled.  I'll jump back in this topic a little later, there's some other changes I need to focus on first.

Just wanted to touch base on this one.

**Comment by a-komarev** [comment]
Any progress on it?

**Comment by eliseumds** [comment]
That would be helpful.

File: croppie.js [hunk 1]
```diff
@@ -54,26 +54,18 @@
...
-                    else
-                        out[key] = obj[key];
-                }
+    // Credits to : Andrew Dupont - http://andrewdupont.net/2009/08/28/deep-extending-objects-in-javascript/
+    function deepExtend(destination, source) {
+        destination = destination || {};
+        for (var property in source) {
+            if (source[property] && source[property].constructor && source[property].constructor === Object) {
+                destination[property] = destination[property] || {};
+                arguments.callee(destination[property], source[property]);
```
Comment (kapouer) [inline croppie.js]: Note that `arguments.callee` is bad, just use `deepExtend` in that case.
See [mdn's doc](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/arguments/callee) for further information.

**Pull request closed without merging**